cli/src/migrate.js

```diff
@@ -0,0 +1,354 @@
+}
+
+function makeBackup() {
+  // shell out to rethinkdb dump
+  const rdbHost = this.options.rdb_host;
+  const rdbPort = this.options.rdb_port;
+
+  if (this.options.skip_backup) {
+    return Promise.resolve();
+  }
+
+  return new Promise((resolve, reject) => {
+    white('Backing up rethinkdb_data directory');
+    const proc = child_process.spawn('rethinkdb', [
+      'dump',
+      '--connect',
+      `${rdbHost}:${rdbPort}`,
+    ]);
+    proc.on('exit', (code) => {
+      if (code === 0) {
+        green(' └── Backup completed');
+        resolve();
+      } else {
+        proc.stderr.setEncoding('utf8');
+        const err = proc.stderr.read();
+        reject(new Error(`rethinkdb dump exited with an error:\n\n${err}`));
+      }
+    });
+  });
+}
+
+function renameUserTables() {
+  // for each table listed in ${project}_internal.tables
+  // rename the table name to the collection name
+  const project = this.options.project_name;
+  return Promise.resolve().then(() => {
+    white('Removing suffix from user tables');
+    return r.db(`${project}_internal`).table('collections')
+      .forEach((collDoc) => r.db('rethinkdb').table('table_config')
+          .filter({ db: project, name: collDoc('table') })
+          .update({ name: collDoc('id') }))
+      .run(this.conn)
+      .then(() => green(' └── Suffixes removed'));
+  });
+}
+
+function moveInternalTables() {
+  // find project_internal
+  // move all tables from ${project}_internal.${table} to ${project}.hz_${table}
+  //   - except for users, don't add hz_prefix, but move its db
+  const project = this.options.project_name;
+  return Promise.resolve().then(() => {
+    white(`Moving internal tables from ${project}_internal to ${project}`);
+    return r.db('rethinkdb').table('table_config')
+      .filter({ db: `${project}_internal` })
+      .update((table) => ({
+        db: project,
+        name: r.branch(
+          table('name').ne('users'),
+          r('hz_').add(table('name')),
+          'users'),
+      })).run(this.conn)
+      .then(() => green(' ├── Internal tables moved'));
+  }).then(() => {
+    // delete project_internal
+    green(` └── Deleting empty "${project}_internal" database`);
+    return r.dbDrop(`${project}_internal`).run(this.conn);
+  });
+}
+
+function renameIndices() {
+  // for each user $table in ${project}
+  //    for each index in ${table}
+  //        parse the old name into array of field names.
+  //        rename to `hz_${JSON.stringify(fields)}`
+  const project = this.options.project_name;
+  return Promise.resolve().then(() => {
+    white('Renaming indices to new JSON format');
+    return r.db(project).tableList().forEach((tableName) => {
+      return r.db(project).table(tableName).indexList().forEach((indexName) => {
+        return r.db(project).table(tableName).indexRename(indexName, rename(indexName));
+      });
+    }).run(this.conn)
+    .then(() => green(' └── Indices renamed.'));
+  });
+
+  function rename(name) {
```

Comment:
It wasn't *necessary* to do this in ReQL.  Less risky to just reuse the old functions.